pkg/services/auth/jwt.go

```diff
@@ -0,0 +1,78 @@
+package auth
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/pkg/errors"
+
+	"github.com/weaveworks/weave-gitops/pkg/gitproviders"
+
+	"k8s.io/apimachinery/pkg/util/rand"
+
+	"github.com/golang-jwt/jwt"
+)
+
+var SecretKey string
+
+const ExpirationTime = time.Minute * 15
+
+var ErrUnauthorizedToken = errors.New("unauthorized token")
+
+// Claims is a custom JWT claims that contains some token information
+type Claims struct {
+	jwt.StandardClaims
+	Provider      gitproviders.GitProviderName `json:"provider"`
+	ProviderToken string                       `json:"provider_token"`
+}
+
+// GenerateJWT generates and signs a new token
+func GenerateJWT(secretKey string, expirationTime time.Duration, providerName gitproviders.GitProviderName, providerToken string) (string, error) {
+	claims := Claims{
+		StandardClaims: jwt.StandardClaims{
+			ExpiresAt: time.Now().Add(expirationTime).Unix(),
+		},
+		Provider:      providerName,
+		ProviderToken: providerToken,
+	}
+
+	token := jwt.NewWithClaims(jwt.SigningMethodHS512, claims)
+	return token.SignedString([]byte(secretKey))
+}
+
+// VerifyJWT verifies the access token string and return a user claim if the token is valid
+func VerifyJWT(secretKey string, accessToken string) (*Claims, error) {
+	token, err := jwt.ParseWithClaims(
+		accessToken,
+		&Claims{},
+		func(token *jwt.Token) (interface{}, error) {
+			_, ok := token.Method.(*jwt.SigningMethodHMAC)
+			if !ok {
+				return nil, fmt.Errorf("unexpected token signing method")
+			}
+
+			return []byte(secretKey), nil
+		},
+	)
+
+	if err != nil {
+		if ve, ok := err.(*jwt.ValidationError); ok {
```

Comment:
Looking at this again, if we have an error we'll have an Unauthorized token anyway, what do you think? should we always return that `ErrUnauthorizedToken` and not check what type of error is that since we only care that the valid or not?

So we would become just:

```
if err != nil {
   return nil, ErrUnauthorizedToken
}
```

Reply:
Good point.  I'm thinking in the user consuming the api, should the user get a specific error to acknowledge that the api  call was done wrong?. I know we don't want to return `ValidationErrorExpired` because that exposes how long ago the token expired (security reasons) .  I guess this comes down to which errors we consider are security issues. Do all errors that `ParseWithClaims` returns would be considered security issues? @jpellizzari @JamWils any thoughts you might have ?

Reply:
> Do all errors that ParseWithClaims returns would be considered security issues?

Probably? This error happens when someone submits a malformed token, which **is** an `Unauthorized` token. I would be okay with returning an `ErrUnauthorizedToken` here.